ghcide/src/Development/IDE/LSP/Server.hs

```diff
@@ -43,7 +44,12 @@ requestHandler m k = LSP.requestHandler m $ \RequestMessage{_method,_id,_params}
       trace x = otTracedHandler "Request" (show _method) $ \sp -> do
         traceWithSpan sp _params
         x
-  writeChan chan $ ReactorRequest (SomeLspId _id) (trace $ LSP.runLspT env $ resp' =<< k ide _params) (LSP.runLspT env . resp' . Left)
+      item = case _method of
+          SInitialize ->
```

Comment:
This is dead code - `handleInit` is the actual initialize request handler. `lsp` will never actually call a `InitializeRequest` handler defined the regular way since initialize is special. Instead it defines its own handler, into which custom logic can be inserted using `doInitialize`

Reply:
In that case it sounds like `InitializeRequest` is already being handled synchronously, is that right?

Reply:
Yes. 

I've also been confused by the out of order progress messages and something non-obvious is going on. It might even be a bug in lsp-test.

I did try to investigate the ghcide code path and I couldn't figure out how it was possible for us to send progress messages before the initialize response.

Reply:
The second part seems easy to explain: we call `shakeOpen` inside the `doInitialize` handler, `shakeOpen` will evaluate the rules and call `kick`, and `kick` may trigger progress reporting.

Reply:
Right, I thought `shakeOpen` was only initializing state.